text/0000-custom-dst.md

```diff
@@ -0,0 +1,279 @@
+- Feature Name: custom\_dst
+- Start Date: 02 March, 2016
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow Rust code to define custom fat pointers, and define slice code in terms of
+these new operators (instead of the current `std::raw::Repr`).
+
+# Motivation
+[motivation]: #motivation
+
+One major missing feature of Rust is custom dynamically sized types; it's
+impossible to implement a slice type in Rust, for example. This has led to
+things like the `std::raw::Repr` trait, a stopgap measure. It's also very
+difficult to represent things like views into 2d slices; for example,
+pixel buffers. In many places, something other than the Rust slice type is
+needed (if you have both a width and a height, or if you have a width, height,
+and stride, or even further).
+
+# Detailed design
+[design]: #detailed-design
+
+The most important addition to the language shall be one new trait:
+
+```rust
+#[lang = "dst"]
+unsafe trait Dst {
+    type Meta: Copy;
+    fn size_of_val(&self) -> usize;
```

Comment:
please add at least a sketches of comments describing what these associated items are supposed to do.

Here is an example of what I am thinking of (just to illustrate that the descriptions can be succinct).

```rust
/// The `Meta` for a type `T` is extra information attached to any pointer or reference to a `T`.
/// In other words, a non-zero sized `T::Meta` means that `&T` is represented as a so-called "fat pointer."
type Meta: Copy;

/// Returns the size, in bytes, of the `T` according to the `Meta` value attached to `self`.
fn size_of_val(&self) -> usize;
```

(I grok that the rest of the RFC is sort of filling in this gap via examples, but as a reader it is a bit frustrating to have no specification, and then have the following paragraph warn about how error-prone implementing this trait is.)